### PR TITLE
(DOCSP-25140) Monitoring Use Cases

### PR DESCRIPTION
**Pull request by biniona-mongodb**
# Pull Request Info

Adds use cases for monitoring features.

[PR Reviewing Guidelines](https://github.com/mongodb/docs-kafka-connector/blob/master/REVIEWING.md)

[JIRA](https://jira.mongodb.org/browse/DOCSP-25140)
[Staging](https://docs-mongodbcom-staging.corp.mongodb.com/kafka-connector/docsworker-xlarge/DOCSP-25140-Monitoring-Use-Cases/monitoring/#use-cases)

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

### CODE REVIEW

File: source/monitoring.txt [hunk 1]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+       encountered when attempting to write to MongoDB.
+     - Use the ``batch-writes-failed`` metric.
+
+   * - You need to know if your MongoDB performance is getting slower
+       over time.
+     - Use the ``getmore-commands-successful-duration-over-<number>-ms``
+       and ``getmore-commands-failed-duration-over-<number>-ms``
+       metrics. Track the change in these metrics over
+       time, and see if the values of the longer-duration metrics are increasing
+       at a faster rate than the values of the shorter-duration metrics.
```
Comment (jordan-smith721) [inline source/monitoring.txt:77]: This last sentence was a bit confusing to me. I had to read over it a couple times to figure out what the "longer-duration metrics" and "shorter-duration-metrics" are. Is that referring to the amount you put in the <number> placeholder in the command?

File: source/monitoring.txt [hunk 2]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+   :widths: 40 60
+
+   * - Use Case
+     - Metrics to Use
+
+   * - You need the number of change stream events your source connector
+       has processed. 
+     - Use the ``records`` metric.
+
+   * - You need to how old the most recent change stream event your connector
```
Comment (jordan-smith721) [inline source/monitoring.txt:108]:
```suggestion
   * - You need to know how old the most recent change stream event your connector
```

File: source/monitoring.txt [hunk 3]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+       processed is.
+     - Use the ``latest-mongodb-time-difference-secs``
+       metric.
+
+   * - You need to know the percentage of records your connector
+       received that it failed to write to {+ak+}.
```
Comment (jordan-smith721) [inline source/monitoring.txt:114]: Issue: This sentence reads a little difficult to me

Suggestion: "You need to know the percentage of records your connector received and failed to write to {+ak+}.
Or maybe "You need to know the percentage of records your connector received but failed to write to {+ak+}."

File: source/monitoring.txt [hunk 4]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+       connector write MongoDB documents to {+ak+}.
+     - Compare the values of the ``in-task-poll-duration-ms`` and
+       ``in-connect-framework-duration-ms`` metrics.
+
+   * - You need to know if your MongoDB performance is getting slower
+       over time.
+     - Use the ``getmore-commands-successful-duration-over-<number>-ms``
+       metrics. Track the change in these metrics over
+       time, and see if the values of the longer-duration metrics are increasing
+       at a faster rate than the values of the shorter-duration metrics.
```
Comment (jordan-smith721) [inline source/monitoring.txt:131]: Same as above - I'm unsure what the "longer-duration" and "shorter-duration" metrics are here

**Review by jordan-smith721 — CHANGES_REQUESTED** [review]
Just a few suggestions and some phrasing that I'm a bit confused on.

File: source/monitoring.txt (hunk 2 above)
Comment (biniona-mongodb) [reply source/monitoring.txt:108]: 🤦

File: source/monitoring.txt (hunk 3 above)
Comment (biniona-mongodb) [reply source/monitoring.txt:114]: Good catch, you are right definitely could be improved. Taking the 2nd suggestion

File: source/monitoring.txt (hunk 1 above)
Comment (biniona-mongodb) [reply source/monitoring.txt:77]: Good catch, this is definitely hard to parse. Had trouble thinking of this, but probably still more room for improvement. I'll think about this more and get something back to you.

Comment (biniona-mongodb) [reply source/monitoring.txt]: Split this into two sentences as follows:

Track the change in these metrics over time. See if the number of longer running commands is increasing at a faster rate than the number of the shorter running commands.

**Review by jordan-smith721 — APPROVED** [review]
LGTM

File: source/monitoring.txt [hunk 5]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
 
 .. _kafka-monitoring-information:
 
+Use Cases
+---------
+
+This section describes use cases for monitoring MongoDB source and sink
+connectors.
```
Comment (katcharov) [inline source/monitoring.txt:42]: It seems that users would have questions of the forms:

1. How much in total/at present?
2. How much over a span of time?
3. Is the rate changing?
4. What is the value compared to some other metric?

It might be worth explicitly pointing out that we track only the first, and the user is expected to compute or visually determine the remaining. I'm not sure whether to suggest enumerating these different types as I have done above, but the user might be surprised that they cannot immediately see metrics like "rate over the last minute" or "% of time spent in phase X", and are expected to compute them in some way (it wasn't immediately obvious to me when implementing this that we would not be offering such metrics).

Examples:

1. Have I been receiving all records?
2. Is there a high constant rate of errors? How many records, histogram by hours of the day?
3. Is the error rate increasing? Are we falling further and further behind the origin?
4. Is the problem in MongoDB or Kafka, and if in MongoDB, which phase is causing a problem?

(I don't suggest that we do or do not add these examples, they are meant to be illustrative.)

File: source/monitoring.txt [hunk 6]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+     - Metrics to Use
+
+   * - You need to know the total number of records your connector
+       wrote to MongoDB.
+     - Use the ``records`` metric.
+
+   * - You need to know how long the interval of time is between
+       when a record arrives in a Kafka topic and when your connector writes
+       that record to MongoDB.
+     - Use the ``latest-kafka-time-difference-ms`` metric.
```
Comment (katcharov) [inline source/monitoring.txt:65]: This ultimately doesn't change much for the implied use case, but this interval of time is between the writing of the record to Kafka, and receiving it. It does not include the additional delay in writing that record to MongoDB.

File: source/monitoring.txt [hunk 7]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+     - Use the ``records`` metric.
+
+   * - You need to know how old the most recent change stream event your connector
+       processed is.
+     - Use the ``latest-mongodb-time-difference-secs``
+       metric.
+
+   * - You need to know the percentage of records your connector
+       received but failed to write to {+ak+}.
+     - Use the ``records-filtered`` and ``records`` metrics.
```
Comment (katcharov) [inline source/monitoring.txt:115]: The `records-filtered` metric does not imply failure. The user can configure things such that certain records are intentionally not written.

To determine implied failure, the user should take the records metric, and subtract the filtered and acknowledged metrics, and the result is the number of records that were not yet (which is fine) or never will be (due to failure) handled by the Kafka connector. (There is no explicit failure metric, because the framework does not explicitly report failures.)

File: source/monitoring.txt [hunk 8]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+
+   * - You need to know the average size of the documents your connector
+       has processed.
+     - Use the ``mongodb-bytes-read`` and ``records`` metrics.
```
Comment (katcharov) [inline source/monitoring.txt:119]: Maybe explicitly say to divide bytes by records (rather than "use"). 

This is also a place where users might care about a span of time, which I worried might be confusing (as you explain elsewhere, one needs to perform T2-T1 to get the number over a duration, and then perform the math.)

File: source/monitoring.txt [hunk 9]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+
+   * - You need to find a bottleneck in how {+kc+} and your MongoDB source
+       connector write MongoDB documents to {+ak+}.
+     - Compare the values of the ``in-task-poll-duration-ms`` and
+       ``in-connect-framework-duration-ms`` metrics.
+
+   * - You need to know if your MongoDB performance is getting slower
+       over time.
```
Comment (katcharov) [inline source/monitoring.txt:127]: The slowdown can be (and initially, probably should be) inferred from in-task-poll-duration-ms metric as well.

The specific reason we have these metrics is to catch outliers. For example, you might have a rate of 100 commands per 100 seconds because each command takes exactly one second, or because 99 commands take a few ms, but then one command consumes the remaining time. By observing the "over-N" metric, you can differentiate these two cases.

**Review by katcharov — CHANGES_REQUESTED** [review]

File: source/monitoring.txt (hunk 5 above)
Comment (biniona-mongodb) [reply source/monitoring.txt:42]: Sounds good, agree it could be useful to explicitly state what types of metrics the connector provides (counts and most recent occurrence) and when you must compute something (anything that cannot be expressed as a count or most recent occurrence).

Added a subsection explaining this and link to it from use cases subsection.

File: source/monitoring.txt [hunk 10]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+     - Use the ``latest-kafka-time-difference-ms`` metric.
+
+   * - You need to know the total number of write errors your connector
+       encountered when attempting to write to MongoDB.
+     - Use the ``batch-writes-failed`` metric.
+
+   * - You need to know if your MongoDB performance is getting slower
+       over time.
+     - Use the ``getmore-commands-successful-duration-over-<number>-ms``
+       and ``getmore-commands-failed-duration-over-<number>-ms``
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/monitoring.txt]: Added initial + processing commands to this list.

Not sure if there is a natural place to put the information in the 3rd paragraph of your suggestion, but I'll keep it in mind throughout edits.

File: source/monitoring.txt (hunk 6 above)
Comment (biniona-mongodb) [reply source/monitoring.txt:72]: Clarified that this is between when a record arrives on topic and when a connector received a record (does not include writing the record).

File: source/monitoring.txt [hunk 11]
```diff
@@ -35,6 +35,104 @@ To learn more about logging in the connector, see the
...
+
+   * - Use Case
+     - Metrics to Use
+
+   * - You need the number of change stream events your source connector
+       has processed. 
+     - Use the ``records`` metric.
+
+   * - You need to know how old the most recent change stream event your connector
+       processed is.
```
[the remark this answers is not in the document]
Comment (biniona-mongodb) [reply source/monitoring.txt:119]: Updated description to "a component of your pipeline is falling behind" to highlight the value of this metric and moved this item to the top of the table for both source and sink.

File: source/monitoring.txt (hunk 8 above)
Comment (biniona-mongodb) [reply source/monitoring.txt]: Explicitly stated formula reader should use and linked to additional admonition later in page

File: source/monitoring.txt (hunk 9 above)
Comment (biniona-mongodb) [reply source/monitoring.txt:154]: Added information about using the `in-task-poll/put-duration-ms` to initally diagnose a slowdown, and updated langauge to suggest that `over-n` language should be used for further diagnosing any issues.

File: source/monitoring.txt [hunk 12]
```diff
@@ -180,6 +213,28 @@ To learn how to enable JMX for a containerized {+kc+}
...
+- The number of times an event has occurred in total for a connector task
+- The value related to the most recent occurrence of an event
+
+For some use cases, you must perform extra computations with the
+metrics the connector provides. For example, you must compute the
+following values from provided metrics:
+
+- The rate of change of a metric
+- The value of a metric over a span of time
+- How does one metric compare to another metric
```
Comment (katcharov) [inline source/monitoring.txt:233]: Nit: "How one metric compares to another metric" ? (Something seems different about the phrasing of this 3rd item compared to the prior two)

**Review by katcharov — APPROVED** [review]
